test/model/store_test.dart

```diff
@@ -1127,6 +1210,46 @@ void main() {
       // Reload never succeeds and there are no unhandled errors.
       check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
     }));
+
+    test('new store is not loaded, gets InitialSnapshot with ancient server version', () => awaitFakeAsync((async) async {
+      final json = eg.initialSnapshot(zulipFeatureLevel: eg.ancientZulipFeatureLevel).toJson();
+      await prepareReload(async, prepareRegisterQueueResponse: (connection) {
+        connection.prepare(
+          delay: Duration(seconds: 1),
+          json: json);
+      });
+
+      async.elapse(const Duration(seconds: 1));
+      check(globalStore.takeDoRemoveAccountCalls()).single.equals(eg.selfAccount.id);
+
+      async.elapse(TestGlobalStore.removeAccountDuration);
+      check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
+
+      async.flushTimers();
+      // Reload never succeeds and there are no unhandled errors.
+      check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
+    }));
+
+    test('new store is not loaded, gets malformed response with ancient server version', () => awaitFakeAsync((async) async {
+      final json = eg.initialSnapshot(zulipFeatureLevel: eg.ancientZulipFeatureLevel).toJson();
+      json['realm_emoji'] = 123;
+      check(() => InitialSnapshot.fromJson(json)).throws<void>();
+      await prepareReload(async, prepareRegisterQueueResponse: (connection) {
+        connection.prepare(
+          delay: Duration(seconds: 1),
+          json: json);
+      });
+
+      async.elapse(const Duration(seconds: 1));
+      check(globalStore.takeDoRemoveAccountCalls()).single.equals(eg.selfAccount.id);
+
+      async.elapse(TestGlobalStore.removeAccountDuration);
+      check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
+
+      async.flushTimers();
+      // Reload never succeeds and there are no unhandled errors.
+      check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
+    }));
```

Comment:
Other than these tests, I think it would be valuable to have an end-to-end widget test that checks the error dialog when connecting to an ancient server, but I feel like it can be difficult to set up.

Reply:
I agree. We'd need change `TestZulipBinding` so that it could work with a `UpdateMachineTestGlobalStore`, and I'm not sure how best to do that.

The `TestZulipBinding.globalStore` getter has a `TestGlobalStore` return type, and `UpdateMachineTestGlobalStore` doesn't extend `TestGlobalStore`. Both classes extend `GlobalStore`, but that doesn't make sense as a return type; it would break a lot of callers that use things like `TestGlobalStore.add`.